### PR TITLE
[INFRA/RULES] Layered rules implemented for IIGO conditional rules

### PR DESCRIPTION
**Pull request by Neelesh99**
# Summary

Newly conceptualised rules for IIGO conditionals requires an extension to the rules infrastructure. There should be no regression impacts and this is tested. Linked rules allow for a parent rule to conditionally execute a child rule (kind of like an if statement).

## Test Plan

Added tests

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/common/rules/rulematrix.go [hunk 1]
```diff
@@ -4,10 +4,27 @@ import (
...
+const (
+	ParentFailAutoRulePass LinkTypeOption = iota
+	NoLink
+)
+
+// RuleLink provides a containerised package for all linked rules
+type RuleLink struct {
+	Linked     bool
+	LinkType   LinkTypeOption
+	LinkedRule string
```
Comment (Spruudis) [inline internal/common/rules/rulematrix.go:21]: Should this contain a RuleMatrix instead? How would you go about changing the linked rule? Do we still have a rulesInPlay matrix and if so, do both of the linked and the mother rule need to be in play for it to be evaluated properly?

Comment (Neelesh99) [reply internal/common/rules/rulematrix.go:21]: I think we should keep it as a string reference to the rule name, to ensure minimal regression impacts on voting and evaluation. Yes both the linked and mother rule needs to be in play, this makes sense right, we can vote to modify or remove either independent from each other. We don't need to modify any of our procedures but we still get the intended effect.

File: internal/common/rules/basicruleevaluator.go [hunk 2]
```diff
@@ -167,3 +167,32 @@ func BasicRealValuedRuleEvaluator(ruleName string) (bool, float64, error) {
...
+			if link.LinkType == ParentFailAutoRulePass {
+				childRule := link.LinkedRule
+				parentPass, parentErr := BasicBooleanRuleEvaluator(ruleName)
+				if parentErr != nil {
+					return false, errors.Errorf("Parent Rule errored out wiht : %v", parentErr)
+				}
+				if !parentPass {
+					return true, nil
+				}
+				childPass, childErr := BasicBooleanRuleEvaluator(childRule)
```
Comment (Spruudis) [inline internal/common/rules/basicruleevaluator.go]: Here we evaluate the child rule even if it is not in play, is that right or am I misunderstanding?

Comment (Neelesh99) [reply internal/common/rules/basicruleevaluator.go]: Yes, all the rule evaluators feed off the available rules cache to allow for islands to look at prospective rules. It's up to us to check whether the rules are in play before we call this.

File: internal/common/rules/basicruleevaluator_test.go [hunk 3]
```diff
@@ -43,12 +54,38 @@ func registerNewRealValuedRule(t *testing.T) {
...
+
+	v := []float64{1, 0, 0, 0, -4, 0, -1, -1, 0, 2, 0, 0, 0, 0, 2, 0, 0, 1, 0, -1}
+	CoreMatrix := mat.NewDense(4, 5, v)
+	aux := []float64{1, 1, 3, 0}
+	AuxiliaryVector := mat.NewVecDense(4, aux)
+
+	_, ruleError := RegisterNewRule(name, reqVar, *CoreMatrix, *AuxiliaryVector, false, RuleLink{
+		Linked:     true,
+		LinkType:   ParentFailAutoRulePass,
+		LinkedRule: "Kinda Complicated Rule",
```
Comment (NGovani) [inline internal/common/rules/basicruleevaluator_test.go:86]: Where is this rule defined

File: internal/common/rules/rulematrix.go [hunk 4]
```diff
@@ -4,10 +4,27 @@ import (
 	"gonum.org/v1/gonum/mat"
 )
 
+// LinkTypeOption gives an enumerated type for the various link types available for rules
+type LinkTypeOption int
+
+const (
```
Comment (NGovani) [inline internal/common/rules/rulematrix.go:10]: An explanation of what these mean would be nice

**Review by NGovani — CHANGES_REQUESTED** [review]

File: internal/common/rules/basicruleevaluator_test.go (hunk 3 above)
Comment (Neelesh99) [reply internal/common/rules/basicruleevaluator_test.go:86]: BasicRuleEvaluator feeds off whatever is in the AvailableRulesCache, so this rule is a default entry in there, it was a convenient one to use ;) Defined on from line 16 on globalruleregistration.go

File: internal/common/rules/rulematrix.go (hunk 4 above)
Comment (Neelesh99) [reply internal/common/rules/rulematrix.go:10]: Sure added in latest commit

**Review by NGovani — APPROVED** [review]

**Comment by NGovani** [comment]
I'm a bit concerned that isn't used yet, so we don't know how its gonna affect end to end test. But I assume that's coming in a future PR it should be fine to merge

**Pull request merged**